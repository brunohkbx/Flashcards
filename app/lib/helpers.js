import { Platform } from 'react-native';

export const waitFor = ms => {
  return new Promise(resolve => setTimeout(resolve, ms));
};

export const getOS = () => {
  return Platform.OS;
};
