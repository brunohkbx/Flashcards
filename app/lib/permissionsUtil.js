import { Permissions } from 'expo';

export const obtainNotificationPermission = async () => {
  const { status: existingStatus } = await Permissions.getAsync(
    Permissions.NOTIFICATIONS
  );
  let finalStatus = existingStatus;

  if (existingStatus !== 'granted') {
    const { status } = await Permissions.askAsync(Permissions.NOTIFICATIONS);

    finalStatus = status;
  }

  return finalStatus;
};
