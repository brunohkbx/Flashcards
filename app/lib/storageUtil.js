import { AsyncStorage } from 'react-native';
import { pickBy, assign } from 'lodash/object';

const DECK_STORAGE_KEY = 'flashcards:deck';

export const getDecks = () => {
  return AsyncStorage.getItem(DECK_STORAGE_KEY);
}

export const createDeck = deck => {
  return AsyncStorage.mergeItem(DECK_STORAGE_KEY, JSON.stringify(deck));
}

export const deleteDeck = id => {
  return getDecks()
    .then(decks => JSON.parse(decks))
    .then(decks => pickBy(decks, deck => deck.id !== id))
    .then(decks => AsyncStorage.setItem(DECK_STORAGE_KEY, JSON.stringify(decks)));
}

export const editDeck = deck => {
  return getDecks()
    .then(decks => JSON.parse(decks))
    .then(decks => assign(decks, deck))
    .then(decks => AsyncStorage.setItem(
      DECK_STORAGE_KEY, JSON.stringify(decks)
    ));
}
