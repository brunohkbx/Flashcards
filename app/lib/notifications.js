import { Notifications } from 'expo';
import * as storageUtil from '../lib/storageUtil';

export const clearLocalNotification = () => {
  return storageUtil.unscheduleNotification()
    .then(Notifications.cancelAllScheduledNotificationsAsync);
};

export const scheduleLocalNotification = () => {
  return storageUtil.isScheduledNotification()
    .then(JSON.parse)
    .then(data => {
      if (data === null) {
        Notifications.cancelAllScheduledNotificationsAsync();

        let tomorrow = new Date();
        tomorrow.setDate(tomorrow.getDate() + 1);
        tomorrow.setHours(18);
        tomorrow.setMinutes(30);

        Notifications.scheduleLocalNotificationAsync(
          {
            title: 'Haven\'t seen you in a while',
            body: 'ðŸ‘‹ Don\'t forget to take your quiz today!',
            ios: {
              sound: true
            },
            android: {
              sound: true,
              priority: 'high',
              sticky: false,
              vibrate: true
            }
          },
          {
            time: tomorrow,
            repeat: 'day',
          }
        );

        storageUtil.scheduleNotification();
      }
    });
};
