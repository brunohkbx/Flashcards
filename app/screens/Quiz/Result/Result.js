import React, { Component } from 'react';
import { Image } from 'react-native';
import PropTypes from 'prop-types';
import { Colors, Headline, Subheading, Button } from 'react-native-paper';
import CenteredContainer from '../../../components/CenteredContainer';
import Avatar from '../../../components/Avatar';
import { clearLocalNotification, scheduleLocalNotification } from '../../../lib/notifications';

class Result extends Component {

  scheduleNotification = () => {
    return clearLocalNotification().then(scheduleLocalNotification);
  }

  componentDidMount() {
    const { receiveNotifications } = this.props;

    if (receiveNotifications)
      this.scheduleNotification();
  }

  render() {
    const {
      score,
      totalQuestions,
      onRestartQuiz
    } = this.props;

    return (
      <CenteredContainer>
        <Avatar
          width={100}
          height={100}
          borderRadius={50}
          backgroundColor={Colors.white}
        >
          <Image
            source={require('../../../assets/images/clipboard.png')}
            style={{ width: 64, height: 64, backgroundColor: Colors.white }}
          />
        </Avatar>
        <Headline>Congratulations, you're done</Headline>
        <Subheading>
          You've answered {score} questions correctly out of {totalQuestions}
        </Subheading>
        <Button
          mode="contained"
          primary
          style={{ marginTop: 10}}
          onPress={onRestartQuiz}
        >
          Restart Quiz
        </Button>
      </CenteredContainer>
    );
  }
}

Result.propTypes = {
  score: PropTypes.number.isRequired,
  totalQuestions: PropTypes.number.isRequired,
  onRestartQuiz: PropTypes.func.isRequired,
  receiveNotifications: PropTypes.bool
};

Result.defaultProps = {
  receiveNotifications: false
};

export default Result;
