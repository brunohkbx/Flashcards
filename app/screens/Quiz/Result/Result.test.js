import React from 'react';
import { shallow } from 'enzyme';
import Result from './Result';
import * as notificationUtil from '../../../lib/notifications';

describe('Result', () => {
  const setup = propOverrides => {
    const defaultProps = Object.assign({
      score: 3,
      totalQuestions: 10,
      onRestartQuiz: jest.fn()
    }, propOverrides);

    const wrapper = shallow(<Result {...defaultProps} />);
    const wrapperInstance = wrapper.instance();

    return { wrapper, wrapperInstance };
  };

  it('renders properly', () => {
    const { wrapper } = setup();

    expect(wrapper).toMatchSnapshot();
  });

  describe('scheduleNotification', () => {
    it('calls clearLocalNotification and then scheduleLocalNotification', async () => {
      const { wrapperInstance } = setup();
      jest.spyOn(notificationUtil, 'clearLocalNotification');
      jest.spyOn(notificationUtil, 'scheduleLocalNotification');

      await wrapperInstance.scheduleNotification();

      expect(notificationUtil.scheduleLocalNotification).toHaveBeenCalled();
      expect(notificationUtil.clearLocalNotification).toHaveBeenCalled();
    });
  });

  describe('componentDidMount', () => {
    describe('When receiveNotifications is enabled', () => {
      it('calls scheduleNotification', () => {
        const { wrapperInstance } = setup({ receiveNotifications: true });
        jest.spyOn(wrapperInstance, 'scheduleNotification');

        wrapperInstance.componentDidMount();

        expect(wrapperInstance.scheduleNotification).toHaveBeenCalled();
      });
    });
  });
});
