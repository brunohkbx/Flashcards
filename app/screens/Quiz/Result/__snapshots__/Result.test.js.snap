// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Result renders properly 1`] = `
<Styled(Component)>
  <Styled(Component)
    backgroundColor="#ffffff"
    borderRadius={50}
    height={100}
    width={100}
  >
    <Image
      source={1}
      style={
        Object {
          "backgroundColor": "#ffffff",
          "height": 64,
          "width": 64,
        }
      }
    />
  </Styled(Component)>
  <Headline>
    Congratulations, you're done
  </Headline>
  <Subheading>
    You've answered 
    3
     questions correctly out of 
    10
  </Subheading>
  <withTheme(Button)
    mode="contained"
    onPress={[MockFunction]}
    primary={true}
    style={
      Object {
        "marginTop": 10,
      }
    }
  >
    Restart Quiz
  </withTheme(Button)>
</Styled(Component)>
`;
